New task dataset

Thomsch opened the pull request:
Created a new user task dataset based on the overhauled methodology for the user study.

The tasks - and respective one-liners - are now extracted from the **most popular posts** tagged as _bash_ in bashoneliners.com, commandlinefu.com, stackoverflow.com, superuser.com, and unix.stackexchange.com.

This PR adds a new folder _dataset_ at the root containing the scripts used to scrape, clean, filter, and sample the new set of tasks. Additionally, the intermediary datasets are provided along each step.

Thomsch committed:
Add command frequencies

Thomsch committed:
List and count compatible commands with Tellina

Thomsch committed:
Ignore local VSCode settings

Thomsch committed:
Sort compatible utilities

Thomsch committed:
Cleanup compatible commands script

Thomsch committed:
Document dataset sources

Thomsch committed:
Add web scrapers for bashoneliners and commandlinefu

Thomsch committed:
Document pipenv usage to manage Python scripts

Thomsch committed:
Move utilities to dataset/ folder

Thomsch committed:
Delete sample_tasks notebook

The methodology won't use a complicated sampling strategy based on frequencies anymore.

Thomsch committed:
Remove experiment_design folder

Thomsch committed:
Fix section indentation

Thomsch committed:
Add dependency lock

Thomsch committed:
Add raw, unclean data to the repository

Thomsch committed:
Add clean data for bashoneliners and commandlinefu

Thomsch committed:
Fix typo

Thomsch committed:
Dependencies are documented with requirements.txt

Pipenv was causing installation problems with pandas on MacOS for some reason. When installing directly with pip in a pipenv environment, it was fine.

Thomsch committed:
Add compatible commands

Thomsch committed:
Add in and out of scope commands based on NL2Bash

Thomsch committed:
Add bashoneliners filtered data

Thomsch committed:
Add commandlinefu filtered data

Thomsch committed:
Add bash filters used

Thomsch committed:
Add bs4 requirement

Thomsch committed:
Add clean up script for StackExchange

Thomsch committed:
Basic interactive loop through CSV

Thomsch committed:
Add StackExchange title to clean data

Thomsch committed:
Filter out automatically if, for, while, illegal commands

Thomsch committed:
Add top 30 compatible one liners

Thomsch committed:
Fix missing trailing url slash for superuser data

Thomsch committed:
Add top 30 superuser oneliners

Thomsch committed:
Add 2 more manual entries left over

Thomsch committed:
Normalize superuser entries

Thomsch committed:
Add guideline for StackExchange

Thomsch committed:
Add top 30 unix&linux entries

Thomsch committed:
Fix misplaced filtered result

Thomsch committed:
Implement pipeline to read, merge, and sample posts

Thomsch committed:
Add random seed and include title and oneliner in sampled posts

Thomsch committed:
Add title and oneliner for sampled stack exchange posts

Thomsch committed:
Add list of utilities for each command

Thomsch committed:
Analyze utility frequencies of sampled commands

anhnamtran commented on dataset/README.md:
"[...] We verified that [...] have ~at~ **a** score of at least 0."

anhnamtran commented on dataset/web-scrapers/commandlinefu.py:
```diff
@@ -0,0 +1,31 @@
+import requests
+from bs4 import BeautifulSoup
+
+TOTAL_COMMANDS = 14042
+TOP_COMMANDS = 500
```
How are you getting these numbers?

Thomsch replied:
`TOTAL_COMMANDS` is obtained automatically by running the parser until nothing is returned.
`TOP_COMMANDS` is an arbitrary number to obtain the first 500 most popular posts because we don't need all the posts hosted on the website.

anhnamtran replied:
Could you please add those as comments

anhnamtran commented on dataset/utilities/out-of-scope.txt:
```diff
@@ -0,0 +1,8 @@
+1. I/O redirection <, <<
```
This is missing quite a number of cases for "redirection" in general. See https://www.gnu.org/software/bash/manual/html_node/Redirections.html

For all the cases.

Thomsch replied:
I'll make a note for our paper if we describe out-of-scope redirection. I reused the definition from https://github.com/TellinaTool/nl2bash

Thomsch replied:
I added a small note at the beginning of the file. Good catch!

anhnamtran commented on dataset/utilities/out-of-scope.txt:
```diff
@@ -0,0 +1,8 @@
+1. I/O redirection <, <<
+2. Variable assignment =
```
nit: this should probably be more verbose, `a=b` instead of just `=`

anhnamtran commented on dataset/utilities/out-of-scope.txt:
```diff
@@ -0,0 +1,8 @@
+1. I/O redirection <, <<
+2. Variable assignment =
+3. Compound statements:
+- if, for, while, util statements
```
~util~ -> *until*
You're missing a few conditional constructs, see: https://www.gnu.org/software/bash/manual/html_node/Conditional-Constructs.html

Thomsch replied:
Thank you! I didn't these existed. I updated the file for exhaustiveness.

anhnamtran commented on dataset/README.md:
```diff
@@ -0,0 +1,39 @@
+# User Study Dataset
```
Could you have a section explaining what each of the sub-directories are for/brief summaries of the various scripts/jupyter notebook files?

Specifically for the jupyter notebook files, please add descriptions for them. 

A brief summary of the various data files that exists/generated would also be helpful, as there seems to be many.

Thomsch replied:
You're right! Thank you for calling this out lest I forget to document it.

anhnamtran commented on dataset/utilities/out-of-scope.txt:
```diff
@@ -0,0 +1,8 @@
+1. I/O redirection <, <<
```
Not quite sure how verbose this file should be, or what's its use. See the comment I left on the README file.

Thomsch replied:
This file was mostly an easy-to-access reminder so I didn't need to access https://github.com/TellinaTool/nl2bash and its paper every time I wanted to check what was allowed.
Hopefully, the small description I added at the top of the file fixes the confusion.

Thomsch committed:
Add rewrites for tasks 1, 6, 15 with time it took

The rewrites were done by me.

Thomsch committed:
Update description of folder containing dataset

Thomsch committed:
Rewrite tasks

anhnamtran commented on README.md:
```diff
@@ -6,8 +6,9 @@ This repository contains the infrastructure for formally conducting user experim
-- [**client_side**](https://github.com/TellinaTool/tellina_user_experiment/tree/master/server_side): this is the directory that contains the files to be distributed
+- [**client_side**](https://github.com/TellinaTool/tellina_user_experiment/tree/master/server_side): this directory contains the files to be distributed
   to the users of the experiment.
+- [**dataset**](https://github.com/TellinaTool/tellina_user_experiment/tree/master/dataset): this directory contains the dataset of the tasks and the scripts that made it.
```
This might need to be more verbose, as it is a bit confusing to me. Could you elaborate a bit on "the dataset of the tasks and the scripts that made it"? Also I think it's either `data-set` if you want it to technically be one word or just flat out `data set` as two words in the summary.

The directory name can remain `dataset` if it's too hard to update, otherwise, we should probably keep it consistent with it being `data_set`.

Do let me know if I'm missing some context or terminology on why it should be called `dataset` (one word).

Thomsch replied:
I will update this sentence. I find it also not quite right.

Regarding `dataset`, I used one word by force of habit. I found this [useful discussion](https://english.stackexchange.com/questions/2120/which-is-correct-dataset-or-data-set) on the subject. It looks like `data set` is to be preferred for non-technical jargon and academic literature.
Thank you for the feedback

anhnamtran commented on dataset/utilities/compatible-commands.ipynb:
```diff
@@ -0,0 +1,62 @@
+{
```
Can you put a comment at the top on how to compile and view this Jupyter notebook file? 

Assume the same comment applies to all `.ipynb` files.

Thomsch replied:
Thank you for the suggestion. Would it be acceptable to instead put instructions inside `data_set/README.md`? Usually, Jupyter notebooks do not contain meta-instructions for humans such as how to compile it.

anhnamtran replied:
Yeah that should be fine as well.

anhnamtran commented on dataset/README.md:
```diff
@@ -0,0 +1,39 @@
+The query will return a quintuplet of the question's title, id (used to create a link to the question), tags, score, and raw html body. The `INNER JOIN` on `PostTags` and `Tags` enable to filter questions that are tagged with 'bash' and the last `INNER JOIN` on Posts enable to filter questions that have an accepted answer for the top 500 questions ordered by score.
+
+## Manual exploration
+The web scraping and SQL query are based on available webpages for each site:
```
It would be helpful to write python tests for the scrapers, we can discuss on the necessity of this.

Thomsch replied:
I'm not sure it's worth the time investment in this case because they are not very complex and would break immediately if something goes wrong. Additionally, the top posts can change at any time, which would make the tests brittle if this is what we test.

anhnamtran replied:
That's fair enough, I was mainly thinking of just quick correctness checks that verify the format returned is correct, that is if the quintuplet returned is malformed or not. But this can be an enhancement/request rather than a merge blocker.

anhnamtran commented on dataset/README.md:
A small test script to test these `sed` commands would also be helpful. Since they're `Bash`, we can use Bats to test it.

Thomsch replied:
Thank you for the suggestion. I wonder what value we would get out of it? What would be the main advantage?

anhnamtran replied:
Seds and regexes are just generally icky, I suppose reading a test case with what the input/expected output would be helps explains these kinds of stuff. If it's a bit too complicated, maybe just a quick comment on what you're stripping would be nice. 

~Now that I'm looking at the `sed` commands mentioned again, I'm actually not quite sure what they're doing. It looks like they're just matching the URLs without doing any replacement, in which case, why can't we use `grep -e` here?~ My bad, I missed the `s#^#...` part.

anhnamtran reviewed (changes requested):
Some preliminary reviews, since there's a lot here that still needs some more descriptions.

Thomsch committed:
Resample 2 tasks containing `watch` or `kill`

These utilities could not be easily made testable in the context of the user study.

Thomsch committed:
Update rewrites for the two new sampled tasks

Thomsch committed:
Fix truncated one-liner

Thomsch committed:
Add sanity check

Thomsch committed:
Uniform names

Thomsch committed:
Move details on how we collected the stack exchange data

Thomsch committed:
Fix missing elements from out-of-scope.txt

Thomsch committed:
Overhaul of `dataset/` documentation

Thomsch committed:
dataset -> data set

anhnamtran commented on data_set/README.md:
```diff
@@ -0,0 +1,71 @@
+
+## Contents
+- `data-collected/`: Contains the raw posts data parsed from the web.
+- `data-clean/`: Contains the raw posts data that have been cleaned up (referenced as _candidate posts_).
+- `data-filtered/`: Contains the posts that have been filtered from the raw posts data (referenced as _compatible posts_). In most case, only the first top 30 compatible posts have been extracted.
```
nit: can we keep the word separator consistent. Since we're already using `_`, lets use it for these as well.

Thomsch replied:
I agree with you, we should only have one convention.

anhnamtran commented on data_set/README.md:
```diff
@@ -0,0 +1,71 @@
+- `data-clean/`: Contains the raw posts data that have been cleaned up (referenced as _candidate posts_).
+- `data-filtered/`: Contains the posts that have been filtered from the raw posts data (referenced as _compatible posts_). In most case, only the first top 30 compatible posts have been extracted.
+- `taskset/`: Contains the final set of posts that are going to be used in the user experiment (referenced as _sampled posts_).
+- `web-scrapers/`: Contains the web scrapping scripts to extract the posts information from the websites.
```
nit: Same comment about line separator. `_` instead of `-`.

anhnamtran commented on data_set/README.md:
```diff
@@ -0,0 +1,71 @@
+
+## Data Process
+### 1. Post Acquisition
+The posts is collected through via web-scrapping on five source websites: 
```
"[...] collecting ~through~ via web-scrapping [...]" or "[...] collecting through ~via~ web-scrapping [...]" 

Writer's choice, just not both :)

Thomsch replied:
Hahaha. Good catch x)

anhnamtran commented on data_set/README.md:
```diff
@@ -0,0 +1,71 @@
+The data set goes through intermediate steps before its final form. The process is described in [Data Process](#data-process).
+
+## Contents
+- `data-collected/`: Contains the raw posts data parsed from the web.
```
I'm not sure how you would be able to clarify this, but it took me _quite_ a bit to realize that "post" here refers to the posts on the website, not the HTTP `POST` request.

Thomsch replied:
Ah, you have a good point! I added a little disambiguation at the beginning.

anhnamtran commented on data_set/README.md:
```diff
@@ -0,0 +1,71 @@
+
+## Data Process
+### 1. Post Acquisition
+The posts is collected through via web-scrapping on five source websites: 
```
"The posts ~is~ are collected [...]"

Thomsch replied:
Thanks!

anhnamtran commented on data_set/README.md:
```diff
@@ -0,0 +1,71 @@
+### 3. Post Filtering
+The goal of this step is too filter the _candidate posts_ that are compatible with Tellina and our user experiment.
+
+For CommandLineFu and Bash One-Liners, the process uses various RegExps that are stored in `data-filtered/filters.txt`. Each command is called manually to gradually remove incompatible posts from the _candidate posts_.
```
Why isn't `data-filtered/filters.txt` not just a Bash script? Would this add too much complexity?

Thomsch replied:
That's a good question. I contemplated putting it in the root of `data_set/` or in the utilities. 

It's a `.txt` because I used a manual process to filter out out-of-scope commands as I encountered them. Adapting each command as I relearned Bash and what regexp I needed to be more efficient. Thus, I didn't spend the time consolidating them into a script since I had gotten what I wanted.

anhnamtran commented on data_set/clean_stackexchanges.sh:
```diff
@@ -0,0 +1,4 @@
+#!/bin/bash
+python clean_stackexchange.py data-collected/stackoverflow-top500.csv https://stackoverflow.com/questions/
+python clean_stackexchange.py data-collected/unixlinux-top500.csv https://unix.stackexchange.com/questions/
+python clean_stackexchange.py data-collected/superuser-top500.csv https://superuser.com/questions/
```
Do we want to make this extendable? As in, if we wanted to add sites in the future? 

If so, I would suggest using Bash arrays (or just strings as that's more portable across Bash versions), something along the lines of
```
DATA_DIR="data_collected"
SITES="stackoverflow|unix.stackexchange|superuser"
SITE_TRAILER="questions"
IFS='|' for site in "$SITES"; do
    python clean_stackexchange.py "$DATA_DIR"/"${site}-top500.csv" 'https://'"${site}.com/$SITE_TRAILER"
done
```
This way, adding or changing a site should be relatively easy. 

If not, then the current script should be fine.

Thomsch replied:
I love it! I didn't know it was possible to do that in Bash (it would have taken me a lot of time as I'm not fluent in Bash).
I'm integrating it right away. Thank you for the suggestion and the code!

anhnamtran commented on data_set/taskset/commands.csv:
```diff
@@ -0,0 +1,17 @@
+Id, Post,Commands
```
nit: might be a bug in the generator, there's an extra space here.

Thomsch replied:
Thank you for spotting it out. It's actually a manual error. This file is not generated. I didn't automate the process since it was only 16 commands.

Thomsch replied:
(I fixed the spaces though)

anhnamtran commented on data_set/taskset/rewrites.csv:
```diff
@@ -0,0 +1,17 @@
+Id, Rewrite (* Means that the task has been adapted for the context of the experiment.)
```
nit: same as above. I suppose just being consistent would be fine (space after all commas or no space)

anhnamtran commented on data_set/taskset/sanity_check.md:
```diff
@@ -0,0 +1,276 @@
+# Tellina sanity check
+Sanity check to ensure that the current version of Tellina can handle this representative set of tasks.
+
+A set of suggestions were tagged as 'helpful` if it contains the answer or a part of the answer. If the suggestions were nonsensical or unusable to solve the task, it was tagged as 'unhelpful'.
```
Formatting error here. "'helpful\`" should be "'helpful'" (the little back tick)

Thomsch replied:
Oh wow! Thanks a lot for spotting that!

anhnamtran commented on data_set/taskset/sanity_check.md:
```diff
@@ -0,0 +1,276 @@
+- Using original prompts: 8 helpful, 8 unhelpful
+- Adapted prompts: 15 helpful, 1 unhelpful
+
+Tellina has often trouble with vague sentences or ones that have no context, which is why with no modifications, the original prompts do so poorly: They do not represent how programmers would use Tellina. Indeed, with the adapted prompts that are closer to the prompts participants would input in Tellina, we observe much more helpful results.
```
"Tellina ~has often~ often has trouble [...]"

anhnamtran commented on data_set/taskset/sanity_check.md:
```diff
@@ -0,0 +1,276 @@
+Tellina has often trouble with vague sentences or ones that have no context, which is why with no modifications, the original prompts do so poorly: They do not represent how programmers would use Tellina. Indeed, with the adapted prompts that are closer to the prompts participants would input in Tellina, we observe much more helpful results.
+
+# Details
+For each sampled post, we try the original prompt as well as an adaptation that's closer on how Tellina will be used during the user experiment.
```
"[...] an adaptation that's closer ~on~ to how [...]"

anhnamtran commented on data_set/taskset/sanity_check.md:
```diff
@@ -0,0 +1,276 @@
+
+Helpful
+
+### Adapted prompt: Print the lines in foo.txt that do are not in bar.txt
```
"[...] that ~do~ are not in [...]"

anhnamtran commented on data_set/taskset/sanity_check.md:
```diff
@@ -0,0 +1,276 @@
+
+Helpful
+
+## 2. https://www.commandlinefu.com/commands/view/42/display-which-distro-is-installed (Prompt Display which distro is installed) 
```
"[...]-is-installed (~Prompt~ Display [...])"

Thomsch replied:
Thanks!

anhnamtran commented on data_set/taskset/sanity_check.md:
```diff
@@ -0,0 +1,276 @@
+## 2. https://www.commandlinefu.com/commands/view/42/display-which-distro-is-installed (Prompt Display which distro is installed) 
+One-liner: `cat <file>`
+
+### Prompt: Display which distro is installed
+- `find . -name '[__SP__UNK][__SP__UNK][__SP__UNK][__SP__UNK][__SP__UNK][__SP__UNK][__SP__UNK][__SP__UNK][__SP__UNK][__SP__UNK][__SP__UNK][__SP__UNK][__SP__UNK][__SP__UNK][__SP__UNK][__SP__UNK]__SP__UNK'
+
+Unhelpful
+
+### Adapted prompt: Display the content of index.html
```
These are just two different prompts? I'm guess adapted does not necessarily mean "slightly changed so Tellina can understand" but more "if Tellina can't understand it, we're just using something else"? If that's the case you might need to clarify.

Thomsch replied:
That's exactly right. Even if Tellina understands the original prompt which is the title of a post, I check with an adapted prompt that's closer to how someone would solve a particular problem.

I added a description for adapted prompts at the beginning. Thank you for the feedback.

anhnamtran commented on data_set/taskset/sanity_check.md:
```diff
@@ -0,0 +1,276 @@
+
+Helpful (`mkdir -p`)
+
+## 5. https://stackoverflow.com/questions/3137094/
+One-liner: `wc -l <filename>`
+How to count lines in a document?
```
Can we move "How to count lines in a document?" to line 82 to stay consistent with how we've been doing these sections?

Thomsch replied:
Absolutely! This one must have slipped out unnoticed.

anhnamtran commented on data_set/taskset/sanity_check.md:
```diff
@@ -0,0 +1,276 @@
+
+Helpful
+
+## 6. https://superuser.com/questions/482435/
```
Can you put the prompt on this line as well for consistency.

Thomsch replied:
For sure! IDK why it's not there.

anhnamtran commented on data_set/taskset/sanity_check.md:
```diff
@@ -0,0 +1,276 @@
+
+Helpful
+
+## 7. https://stackoverflow.com/questions/4997693/
```
Same as above

anhnamtran commented on data_set/taskset/sanity_check.md:
```diff
@@ -0,0 +1,276 @@
+
+Helpful
+
+## 8. https://unix.stackexchange.com/questions/144208/
```
Uhhh. I guess just try to stay consistent either way. Since the rest of them don't seem to have the prompt in the section header either.

Thomsch replied:
I fixed all the inconsistencies. I thought I did but apparently not. Thank you for calling it out!

anhnamtran reviewed (changes requested):
Some more changes, but it's starting to look really good!

Thomsch committed:
Document web scapers constants

Thomsch committed:
Uniform file names to use underscore instead of dashes

Thomsch committed:
Add disambiguation for the term 'post'

Thomsch committed:
Integrate loop implementatino for clean stack exchange script

This solution has been proposed here: https://github.com/TellinaTool/tellina_user_experiment/pull/16#discussion_r593555989

Thomsch committed:
Underscore fixes

Thomsch committed:
Fix CSV spaces

Thomsch committed:
Explain what is the `prompt` and `adapted prompt`

Thomsch committed:
Fix inconsistencies

Thomsch committed:
Fix typos

Thomsch commented:
Thank you for the thorough feedback! I really appreciate it.

anhnamtran reviewed (approved):
LGTM.

Could you please merge and squash? Please use a descriptive commit message for the merge commit.

Thomsch committed:
Merge branch 'master' into task-dataset
